utilities/write_batch_with_index/write_batch_with_index.cc

```diff
@@ -240,21 +290,32 @@ class BaseDeltaIterator : public Iterator {
   }
 
   void AdvanceDelta() {
-    if (forward_) {
+    if (IsMovingForward()) {
       delta_iterator_->Next();
     } else {
       delta_iterator_->Prev();
     }
   }
   void AdvanceBase() {
-    if (forward_) {
+    if (IsMovingForward()) {
       base_iterator_->Next();
     } else {
       base_iterator_->Prev();
     }
   }
-  bool BaseValid() const { return base_iterator_->Valid(); }
-  bool DeltaValid() const { return delta_iterator_->Valid(); }
+  bool BaseValid() const {
+    // NOTE: we don't need the bounds check on
+    // base_iterator if the base iterator has an
+    // upper_bounds_check already
+    return base_iterator_->Valid() &&
+           (base_iterator_->has_upper_bound()
+                ? true
+                : IsWithinBounds(base_iterator_->key()));
+  }
+  bool DeltaValid() const {
+    return delta_iterator_->Valid() &&
+           IsWithinBounds(delta_iterator_->Entry().key);
+  }
```

Comment:
Yes, that may be a possible optimisation...

In the first pass I wanted to focus on correctness with as little modification to the current approach as possible. As it is tricky to say the least! I would still like to see this PR merged before optimisations are considered. If there is no upper or lower bound set I don't think this PR will have much (if any) impact on current performance.

I suspect few people can have been using bounds with WBWI - otherwise like me, they would likely have noticed the bugs ;-)

I would prefer that later a subsequent PR could be created which would then add such an optimisation for when bounds are in play.